sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala

```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
     DataFrameHolder(
       _sqlContext.internalCreateDataFrame(rows, StructType(StructField("_1", dataType) :: Nil)))
   }
+
+  /**
+   * ::Experimental::
+   *
+   * Pimp my library decorator for tungsten caching of DataFrames.
+   * @since 1.5.1
+   */
+  @Experimental
+  implicit class TungstenCache(df: DataFrame) {
```

Comment:
For code cleanliness, I wonder whether we should move the bulk of the logic out of this class and into its own file.